esx_service/utils/auth_api.py

```diff
@@ -365,11 +365,68 @@ def is_vm_duplicate(vm_list):
 
     return None
 
+def check_default_datastore(datastore_name):
+    """ Check datastore with given name is a valid value for default_datastore """
+    # The valid default_datastore name are:
+    # named datastore existing on the host
+    # hard coded datastore name "__VM_DS"
+    # "__ALL_DS" is not a valid value to set as "default_datastore"
+    if datastore_name == auth_data_const.VM_DS:
+        return None
+    if datastore_name == auth_data_const.ALL_DS:
+        return error_code.generate_error_info(ErrorCode.DEFAULT_DS_NAME_INVALID, datastore_name)
+
+    if not vmdk_utils.validate_datastore(datastore_name):
+        error_info = error_code.generate_error_info(ErrorCode.DS_NOT_EXIST, datastore_name)
+        return error_info
+
+    return None
+
+def set_default_ds_and_create_privilege(tenant, default_datastore, check_existing):
+    """
+        Set "default_datastore" for given tenant and create a full access privilege
+        to "default_datastore" if entry does not exist
+    """
+    logging.debug("set_default_ds_and_create_privilege: tenant=%s default_datastore=%s check_existing=%d",
+                  tenant, default_datastore, check_existing)
+
+    error_info, auth_mgr = get_auth_mgr_object()
+    if error_info:
+        return error_info
+
+    datastore_url = vmdk_utils.get_datastore_url(default_datastore)
+    if check_existing:
+        error_msg, existing_default_ds_url = tenant.get_default_datastore(auth_mgr.conn)
+        if error_msg:
+            error_info = error_code.generate_error_info(ErrorCode.INTERNAL_ERROR, error_msg)
+            return error_info
+
+        # the "default_datastore" to be set is the same as existing "default_datastore" for this tenant
+        if datastore_url == existing_default_ds_url:
+            return None
+
+    error_msg = tenant.set_default_datastore(auth_mgr.conn, datastore_url)
+    if error_msg:
+        error_info = error_code.generate_error_info(ErrorCode.INTERNAL_ERROR, error_msg)
+        return error_info
+
+    # create full access privilege to default_datastore
+    error_info = _tenant_access_add(name=tenant.name,
+                                    datastore=default_datastore,
+                                    allow_create=True)
+    # privilege to default_datastore already exist, no need to create
+    if error_info and error_info.code == ErrorCode.PRIVILEGE_ALREADY_EXIST:
+        logging.info(error_info.msg + " not overwriting the existing privilege")
+        error_info = None
```

Comment:
With this fix, user cannot update the "default_datastore" through "tenant access" command. Instead, he/she need to use "tenant update --default-datastore". Internally, it calls "set_default_datastore" function. If it fails, an error will be returned. There is no "restore", if the change to new "default_datastore" fails, then the old "default_datastore" should keep unchanged.